Step4 로또(2등) PR 드립니다.

develo-pyo opened the pull request:
안녕하세요 강태화 리뷰어님!
퇴근 후 개인시간 할애하여 리뷰하시기 정말 힘드실 것 같습니다.
고생이 많으십니다 👍 

지난번 step3 에서 부족하다고 생각한 부분인
- 일급객체 외부 가공 불가 처리
- 메소드 더 작게 리팩토링
- 원시타입 래퍼 객체 추가
등을 우선적으로 보완하고, step4 를 진행했습니다.

그럼 리뷰 부탁드리겠습니다.
감사합니다 😊

develo-pyo committed:
style: 변수명, 패키지 경로, 들여쓰기 수정

develo-pyo committed:
refactor: 로또번호 래핑 객체로 분리

1.로또번호 래핑객체로 분리
2.로또번호 유효성검사 객체 내에서 수행하도록 변경

develo-pyo committed:
refactor: 불필요 유효성 검사 제거

원시타입 래핑 클래스 내에서 유효성 검사 수행하도록 책임 분리

develo-pyo committed:
docs: step4 요구사항 분석

develo-pyo committed:
feat: 로또등수 enum 구현

1.로또등수 enum 구현
2.enum 테스트 구현

develo-pyo committed:
feat: 보너스 번호 입력 메소드 구현

develo-pyo committed:
feat: 로또 게임시작시 보너스 번호 매개변수 추가

1. 로또 게임시작시 보너스 번호 매개변수 추가
2. 보너스번호 당첨번호와 중복 유효성검사 추가
3. 보너스번호 당첨번호와 중복 유효성검사 테스트 추가

develo-pyo committed:
feat: 로또 적중 갯수 판별 메소드에 보너스번호 체크 로직 추가 및 구현

1. 로또 적중 갯수 판별 메소드에 보너스번호 체크 로직 추가 구현
2. 로또 번호 유효성검사 (6개 체크) 누락 추가
3. 로또 번호 적중 갯수 테스트 구현

develo-pyo committed:
fix: 일급객체 직접 가공 불가하도록 수정

1. 일급객체 직접 가공 불가하도록 수정
2. 불필요한 테스트 제거

develo-pyo committed:
feat: 로또 결과 도출 메소드에 보너스번호 매개변수 추가

develo-pyo committed:
feat: 로또 통계 결과 출력 view 2등 문구 추가

develo-pyo committed:
feat: 로또 수익률 계산시 enum 사용하도록 로직 수정

1.로또 수익률 계산시 enum 사용하도록 로직 수정
2.테스트 수정

develo-pyo committed:
test: 로또 적중 결과 추출 메소드 테스트 구현

wooobo commented on docs/lotto/README.md:
```diff
@@ -46,4 +49,10 @@
+    * 수익률 노출  
+-----
+### step4 추가 요구사항 (2등 구현)
+1. 보너스볼 번호를 입력받는다.
+  * 보너스볼은 지난주 로또 당첨번호와 중복될 수 없다.
+2. 2등 여부 판별 
+  * 지난주 로또 당첨번호와 로또번호가 5개 일치되고 && 보너스볼 번호를 포함할 경우 2등
```
```suggestion
1. 보너스볼 번호를 입력받는다.
 - [] 보너스볼은 지난주 로또 당첨번호와 중복될 수 없다.
2. 2등 여부 판별 
  - [x] 지난주 로또 당첨번호와 로또번호가 5개 일치되고 && 보너스볼 번호를 포함할 경우 2등
```

작업전 미리 작성하고, 완료된부분 하나하나 체크해 나간다면, 작업에 대해 누락되거나 놓치는 부분을 쉽게 파악할 수 있을것 같아요!

그리고, 작업 중 추가적인 이슈가 발생한다면 바로 하지 못하더라도 작업사항에 기입해놓는것도 좋을것 같구요!

develo-pyo replied:
체크리스트로 활용하기 좋은 방법같습니다. 앞으로 참고하여 진행하겠습니다 감사합니다.

wooobo commented on src/main/java/lotto/controller/LottoGame.java:
```diff
@@ -1,12 +1,11 @@
+import java.util.ArrayList;
 import java.util.List;
 
 public class LottoGame {
```
```
 public LottoGame(int inputPayAmount) {
        payAmount = new PayAmount(inputPayAmount);
        printLottoAmount();
        lottos = new Lottos(payAmount.lottoAmount());
        printLottos();
    }
```
[[엘레강트 오브젝트] '생성자에 코드를 넣지 마세요' 적용해보기](https://cozzin.tistory.com/71)

위 블로그 참고해주셔서 한번 고민해보면 좋을것 같아요 :)

develo-pyo replied:
관련 내용 학습 및 수정하도록 하겠습니다.
엘레강트 오브젝트 도서도 꼭 구매하여 읽어보겠습니다.
감사합니다!

develo-pyo replied:
생성자를 정적팩토리메소드를 사용하여 개선하는 방향은 어떻게 생각하시나요?
공유해주신 링크 내용과 비슷한 효과를 볼 수 있고,
좀 더 단순하고 구조적으로 덜 복잡한 방향이라고 판단해서 정적팩토리 메소드를 사용하여 객체생성을 할 수 있도록 처리해봤습니다.
생성자 하나까지 고민하고 만드는게 정말 즐겁고 값진 경험이네요 😃

wooobo replied:
정적팩토리메소드로 생성자 인자를 다양하게 쓸수 있고 역할에 대한 생성자 네이밍을 명시적으로 줄 수 있으므로 장점이 있다고 생각해요.
좋은 방법중 하나라고 생각해요!

wooobo commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,41 +1,55 @@
 
     public Lotto() {
-        lottoNumbers = LottoNumberGenerator.generateLottoNumbers();
+        lottoNumbers = LottoNumberGenerator.generateLottoNumbers().lottoNumbers;
```
[디미터 법칙(Law of Demeter)](https://tecoble.techcourse.co.kr/post/2020-06-02-law-of-demeter/)

참고해주세요~

develo-pyo replied:
생성자 내에서 인스턴스 변수를 생성하기보다 외부에서 생성하여 넘겨주는 방식으로 처리하는게 낫겠군요!
생각해보니 인자값 있는 생성자만 사용하면 될 것 같습니다.
디미터 법칙도 살펴보았고 공부 좀 더 해서 개인개발블로그에 정리해놔야겠네요. 
감사합니다!

wooobo commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,41 +1,60 @@
 
-    public int countCollectNumber(Lotto winningNumbers) {
+    private void validateLottoNumber(List<Integer> lottoNumbers) {
+        InputValidator.validateLottoNumberCount(lottoNumbers.size());
```
예외가 발생되는데, 사용자에게는 메시지가 표시되지 않는것 같아요.
예외를 던지는 만큼 try catch를 통해 사용자에게 보여주면 더 좋을것 같아요

develo-pyo replied:
예외만 던지고 메시지는 콘솔창에서 저만 확인하고 있었네요... 😷 
수정하겠습니다!

wooobo commented on src/main/java/lotto/util/ProfitCalculator.java:
```diff
@@ -1,18 +1,16 @@
-        for (int i = MIN_WINNING_NUM; i < PRIZE_MONEY.length; i++) {
-            profit += lottoResult.getLottoResult(i) * PRIZE_MONEY[i];
+        for (Rank rank : lottoResult.getLottoResult().keySet()) {
+            profit += rank.getWinningMoney() * lottoResult.getLottoResult().getOrDefault(rank, 0);
```
데이터를 외부에서 가져와서 판단하는것이 아닌 
데이터를 가진 객체에게 메시지를 전달해보면 어떨까요?
```suggestion
         profit += lottoResult.profit(rank);
```
```
#LottoResult
public double profit(Rank rank) {
      return rank.winningMoney(lottoResult.getOrDefault(rank, 0));
}
```

[[객체지향 생활체조 원칙] 규칙 9. getter/setter/property를 쓰지 않는다](https://limdingdong.tistory.com/15)

참고해주세요!

develo-pyo replied:
사실 이부분은.. 
LottoResult 객체에 책임을 부여하여 해당 클래스 내에서 수익률을 계산 및 반환할 수 있도록 처음에 구현을 했었는데요. Step3 리뷰에서 리뷰어님이 별도의 객체로 분리하는게 나을 것 같다는 의견을 주셔서 분리했던 내용입니다. 당시에도 사실 별도 객체로 분리하기 보단 객체에 책임을 주는게 나은 방향이 아닌가 싶긴 했는데 일단 리뷰어님 말씀에 따라 진행했습니다 😢  
최초 PR 에서 구조화가 너무 엉성하여 리뷰어님께서 리뷰를 하시다 잘못 말씀해주신걸수도 있고, 제가 잘못이해한걸수도 있겠네요..
다시 객체에 책임을 부여하도록 하겠습니다! 😊

develo-pyo replied:
최대한 객체에 책임을 부여한다고 고민하며 구현한건데,
해당 부분도 객체 내에서 처리할 수 있었겠네요. 
생각이 짧았습니다.
해당부분 수정했습니다. 
감동의 소스리뷰입니다. 감사합니다 🙏

wooobo reviewed (changes requested):
4단계 잘진행해주셨습니다!

아쉬운 마음에 조금 고민해 보시면 좋을것 같은 부분 코멘트 남겨두었습니다.

확인 후 재요청 부탁드립니다 :pray:

develo-pyo committed:
fix: 예외발생시 메시지 출력하도록 개선

develo-pyo committed:
refactor: 생성자내에서 외부 객체의 값을 통해 인스턴스 생성하지 않도록 개선 - 디미터법칙

1. Lotto: 불필요생성자 제거
2. LottoNumberGenerator: 로또번호 생성시 원시타입 리턴하도록 수정
3. LottoNumberGeneratorTest: 테스트 코드 수정
4. Lottos: Lotto생성시 Lotto 인스턴스에 필요한 값을 외부에서 전달할 수 있도록 수정

develo-pyo committed:
refactor: 수익률 계산을 LottoResult 객체 내에서 처리할 수 있도록 책임 부여

1. ProfitCalculator: 객체의 원시타입을 getter 로 읽지 않고 객체에 메시지 전달하여 수익률 가져오도록 개선
2. ProfitCalculatorTest: 테스트코드 수정
3. LottoResult: 수익률 계산 메소드 구현

develo-pyo committed:
refactor: Lotto 생성자 정적팩토리메소드 패턴으로 변경

develo-pyo committed:
refactor: LottoGame 생성자 정적팩토리메소드 패턴으로 변경

wooobo commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -15,12 +13,7 @@ public class Lotto {
-
     public Lotto(List<Integer> lottoNumbers) {
-        validateLottoNumber(lottoNumbers);
         this.lottoNumbers = mapToLotto(lottoNumbers);
```
생성자 주입된 인자에 대한 값을 변경하는 코드를 지양해라는 관점이고, `validate`는 있어야 한다는게 저의 관점이에요.

왜냐하면 `주 생성자` 는 그 객체가 생성되어야 하는 조건을 가지고 있어야 한다고 생각하기 때문이에요.

develo-pyo replied:
이 부분은 고민끝에 정적팩토리 메소드 안쪽으로 옮긴건데, 생성자에 넣는게 더 적합하겠네요.

wooobo commented on src/main/java/lotto/controller/LottoGame.java:
```diff
@@ -1,42 +1,57 @@
-        lottos = new Lottos(payAmount.lottoAmount());
-        printLottos();
+    public LottoGame(PayAmount payAmount, Lottos lottos){
+        this.payAmount = payAmount;
```
주입되는 인자에 대한 손을 대지 않도록 수정해보셨네요 👍

wooobo commented on src/main/java/lotto/view/OutputView.java:
```diff
@@ -32,9 +34,20 @@ public static void printLottoResult(LottoResult lottoResult, double profitRatio)
-            System.out.printf(OUT_MESSAGE_LOTTO_RESULT, i, PRIZE_MONEY[i], lottoResult.getLottoResult(i));
+        Arrays.stream(Rank.values())
+                .sorted(Comparator.comparingInt(Rank::getWinningMoney))
+                .filter(rank -> rank.getCountOfMatch() > 2)
```
```suggestion
                .filter(rank -> rank.is적절한네이밍())
```

데이터를 가지고 있는 객체한테 메시지를 전달해보는것도 좋겠네요 !

develo-pyo replied:
앗.. 수정하겠습니다!

wooobo commented on src/main/java/lotto/domain/Lotto.java:
```diff
@@ -1,41 +1,46 @@
+    public static Lotto of(List<Integer> lottoNumbers){
+        InputValidator.validateLottoNumberCount(lottoNumbers.size());
+        InputValidator.validateDuplicateLottoNumber(lottoNumbers);
+        return new Lotto(lottoNumbers.stream().map(LottoNumber::new).collect(Collectors.toList()));
```
```suggestion
        return new Lotto(lottoNumbers.stream()
          .map(LottoNumber::new)
          .collect(Collectors.toList()));
```

적절한 개행처리 또한 코딩컨벤션일수 있습니다.
- 가독성향상
- 디버깅시 이점

develo-pyo replied:
넵! 적절히 개행하도록 하겠습니다.

wooobo reviewed (approved):
4단계도 잘 진행해주셨습니다~

몇가지 추가적인 코멘트 남겨놓았습니다.
확인부탁드립니다!

그럼 마지막 단계 시작하시죠!
화이팅입니다 👍
